### DISCUSSION TITLE
MIDI Playback

### DISCUSSION THREAD

**Discussion by Treata11** [orchetect/swift-midi#192] [Q&A] [answered]
I wanted to know wether if this package has the capability of **playing** a MIDI sequence?
Using AVFoundation's `AVMIDIPlayer` you can achieve the very similar thing:
```Swift
    func playMidiFile(at url: URL?) {
        guard let midiURL = url else { return }
        
        do {
            // Create an AVMIDIPlayer instance with the specified URL
            player = try AVMIDIPlayer(contentsOf: midiURL, soundBankURL: nil)
            player.prepareToPlay()
            player.play() {
            }
        } catch {
            print("Failed to play MIDI file: \n\(error)")
        }
    }
    
    func stopPlayback() {
        player.stop()
    }

    var currentPosition: TimeInterval? {
        get { player.currentPosition }
        set { player.currentPosition = newValue! }
    }
    var duration: TimeInterval? {
        player.duration
    }
```
This is just a sample code of what I'm trying to achieve, but unfortunately there are some strange troubles accessing properties & methods other than the `play()` method. It fails to return any values at all & it even fails at stopping the playback once started, the program simply crashes ...
But the issues with the sample code is not the topic of discussion.
I just need to know how to:

1. play a MIDI 
2. stop the playback
3. access the currentPosition in Seconds

using MIDIKit modules.

**Comment by orchetect** [comment, marked as answer]
MIDIKit doesn't offer any playback/sequencer abstractions (yet). The goal is to build a generalized module before the 1.0.0 release that simply provides an event callback with start and stop functions. It would not produce audio or embed any MIDI synthesis of any kind, as MIDIKit deals strictly in MIDI event types and I/O and not audio. The added MIDI synthesis portion could be added in an example project at some point once the base sequencer is built, or better yet it could be done in another package such as AudioKit.

[AudioKit](https://github.com/AudioKit) V6 (in development) has adopted MIDIKit and it has playback/sequencer features. However if I recall it's still being actively developed but will be in before release.

One of the reasons AudioKit does not have fully fledged MIDI playback currently is that AudioKit used to rely on AVFoundation which has been the source of a lot of issues. But V6 is factoring it out almost entirely which means a lot of the convenience classes that were provided by Apple in AVFoundation need to be rewritten by hand which takes time.

**Reply by Treata11** [reply]
> AVFoundation which has been the source of a lot of issues 

Couldn't agree more! AVFoundation caused a lot of frustration for me. It was just pure pain :_)

I appreciate your time reading my issue & thanks for the great explanation.
Good luck to you with your project.

**Reply by Treata11** [reply]
Hi again @orchetect 
Sorry I bother you again with the same question.
**AudioKit** hasn't still developed a MIDI sequencer. Of course as you mentioned it's a time consuming process since AVFoundation been problematic ...
I'm a little short on time & wanted to seek your advice. What is the ultimate MIDI sequencer currently on apple platforms?
I saw some `CoreAudio` minimalist sample codes for this purpose that needed proper handling since they were outdated ... And I heard that It's possible to manage the MIDI playback using `CoreMIDI` APIs (not 100% sure about that) & also there are projects like [MIKMIDI](https://github.com/mixedinkey-opensource/MIKMIDI) ... 
I'd appreciate your opinion on this topic.

**Reply by orchetect** [reply]
I wish I could help more, however I don't have any immediate recommendations or advice in that regard. There are no concrete timelines for when additional features may be developed for our libraries.
